File: src/main/java/com/example/StudentManagementAPI/service/StudentService.java
package com.example.StudentManagementAPI.service;

import com.example.StudentManagementAPI.entity.Student;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.List;

public interface StudentService {


    // ================================
    // ADD STUDENT
    // ================================

    Student addStudent(Student student);


    // ================================
    // GET ALL STUDENTS
    // ================================

    Page<Student> getAllStudents(Pageable pageable);


    // ================================
    // GET STUDENT BY ID
    // ================================

    Student getStudentById(Integer id);


    // ================================
    // UPDATE STUDENT
    // ================================

    Student updateStudent(
            Integer id,
            Student student
    );


    // ================================
    // DELETE STUDENT
    // ================================

    void deleteStudent(Integer id);


    // ================================
    // SEARCH BY NAME
    // ================================

    List<Student> searchByName(String name);


    // ================================
    // SEARCH BY EMAIL
    // ================================

    List<Student> searchByEmail(String email);


    // ================================
    // FILTER BY DEPARTMENT
    // ================================

    List<Student> filterByDepartment(
            String department
    );


    // ================================
    // FILTER BY CITY
    // ================================

    List<Student> filterByCity(
            String city
    );


    // ================================
    // SEARCH + PAGINATION + SORTING
    // ================================

    Page<Student> getStudents(

            String search,

            int page,

            int size,

            String sortBy,

            String direction

    );

}

File: frontend/src/services/StudentService.js
import axios from "axios";

const API_URL = "http://localhost:8080/students";


// Get JWT token and attach it to request
const getAuthHeaders = () => {

    const token = localStorage.getItem("token");

    return {
        headers: {
            Authorization: `Bearer ${token}`
        }
    };

};


// ===============================
// GET ALL STUDENTS
// ===============================

export const getStudents = () => {

    return axios.get(
        API_URL,
        getAuthHeaders()
    );

};


// ===============================
// GET STUDENT BY ID
// ===============================

export const getStudentById = (id) => {

    return axios.get(
        `${API_URL}/${id}`,
        getAuthHeaders()
    );

};


// ===============================
// ADD STUDENT
// ===============================

export const addStudent = (student) => {

    return axios.post(
        API_URL,
        student,
        getAuthHeaders()
    );

};


// ===============================
// UPDATE STUDENT
// ===============================

export const updateStudent = (id, student) => {

    return axios.put(
        `${API_URL}/${id}`,
        student,
        getAuthHeaders()
    );

};


// ===============================
// DELETE STUDENT
// ===============================

export const deleteStudent = (id) => {

    return axios.delete(
        `${API_URL}/${id}`,
        getAuthHeaders()
    );

};